### PR TITLE
feat(cognito): allow to set read and write attributes in Cognito UserPoolClient

### LINKED ISSUE #7407: Allow to set read and write attributes in Cognito UserPoolClient
When a client of a Cognito User Pool is created, the developer can (and should) specify which user attributes the client will be able to read and/or write.

This is not currently implemented in Cognito.

### Use Case

Often you want different clients to be able to make different operations on user attributes.
For example, there might be a `customerPlan` attribute for every user but only some clients (i.e. those used by the back office) are allowed to change its value.

Or you may want to allow a specific client (e.g. the one used by the web app) to be able to read the `email` attribute but not the `

### Proposed Solution

The proposed solution is to allow to set the list of read attributes and write attributes to the props for _userPoolClient_.

As example
```typescript
const pool = new UserPool(stack, 'Pool');
pool.addClient('MyClient', {
  readAttributes: ['email', 'phoneNumber', 'custom:customerPlan', 'custom:isActive']
  writeAttributes: ['phoneNumber', 'custom:isActive']
});
```

I think there are two flaws with this approach:
 1. having to type the attribute names is error prone
 2. having to prepend `custom:` to all custom attributes can be easily forgot

The solution for both problems could be the following:

```typescript
const pool = new UserPool(stack, 'Pool');
pool.addClient('MyClient', {
  readAttributes: {
    // all cognito attributes are automatically available as boolean values
    email: true,
    phoneNumber: true,
    // all custom attributes can be set with an array without the need to prepend them with `custom:`
    custom: ['customerPlan', 'isActive'],
  },
  writeAttributes: {
    // all cognito attributes are available as boolean values
    email: false,
    phoneNumber: true,
    // all custom attributes can be set with an array without the need to prepend them with `custom:`
    custom: ['isActive'],
  },
});
```

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by tom139**
### Commit Message

feat(cognito): allow to set read and write attributes in Cognito UserPoolClient

Add ability to specify which attributes each App Client can read or write.

closes #7407

### End Commit Message
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:403]:
```suggestion
Clients can be configured so that only specific user attributes can be read or modified by it.
```

File: packages/@aws-cdk/aws-cognito/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:404]:
```suggestion
The default behavior is to allow read and write all available attributes, but it is often not the best option.
```

File: packages/@aws-cdk/aws-cognito/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:406]:
```suggestion
For example, consider a webapp that contains a custom attribute `customerPlan` to store the user's pricing plan
where only some clients, such as the administrators, should be allowed to change this value. Read more about [Attribute
Permissions and Scopes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-attribute-permissions-and-scopes).
```

File: packages/@aws-cdk/aws-cognito/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:417]: Can you modify this example to align with the pricing plan example you have above?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:208]: `StandardAttribute` is currently not exported. Instead of using this, can we reuse `RequiredAttributes` here?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:183]:
```suggestion
   * @default - all attributes are readable
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:194]:
```suggestion
   * @default - all attributes are writable
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:216]:
```suggestion
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:192]:
```suggestion
   * The attributes this client will be able to write.
   * This should be used to restrict the attributes the client will be able to write,
   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-attribute-permissions-and-scopes
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:181]:
```suggestion
   * The attributes this client will be able to read.
   * This should be used to restrict the attributes the client will be
   * able to read.
   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-attribute-permissions-and-scopes
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:208]: I looked into the `RequiredAttributes` enum but the `email_verified` and `phone_verified` attributes are missing.
Do you think it would be better to refactor `StandardAttribute` to be an interface extending `RequiredAttributes`?

```ts
export interface StandardAttributes extends RequiredAttributes {
  emailVerified?: boolean
  phoneVerified?: boolean
  allStandardAttributes?: boolean // this could be added to make easier to flag all attributes
}
```

What do you think? Should I refactor to the `interface`?

File: packages/@aws-cdk/aws-cognito/README.md
Comment (tom139) [inline packages/@aws-cdk/aws-cognito/README.md:426]: I've added those lines but I'm not sure it's easy to understand what i mean. @nija-at could you please give me a feedback?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:206]: Extending `RequiredAttributes` and adding `emailVerified` and `phoneVerified` to it sounds like a good idea. However, it can't be named `StandardAttributes` because they're not under [the list of standard attributes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#cognito-user-pools-standard-attributes). How about using a different name for this, such as, `UserPoolClientAttributes`?

> `allStandardAttributes?: boolean // this could be added to make easier to flag all attributes`

As for this, we could do create a method along the lines of `UserPoolClientAttributes.allStandardAttributes()`.

Take a look at how we've done something similar with `AccessLog` - https://github.com/aws/aws-cdk/blob/40fa93a22ffbdf18b0563d1cef63bbf5814dcc3f/packages/%40aws-cdk/aws-apigateway/lib/access-log.ts#L484-L519

File: packages/@aws-cdk/aws-cognito/README.md
Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/README.md:426]: This applies only for custom attributes, no? Because I thought standard attributes are always available.

**Review by nija-at — DISMISSED** [review]
Please address existing comments.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:206]: I'm sorry for the delay.

What do you think about creating an interface to hold all attributes to be set for a client (both standard and custom)
```ts
export interface IUserPoolClientAttributes extends RequiredAttributes {
  readonly emailVerified?: boolean;
  readonly phoneVerified?: boolean;
  // this will allow to set custom and standard attributes from this interface
  [customAttribute: string]: boolean | undefined;
}
```

and a class for conveniently create instances of that interface
```ts
export class UserPoolClientAttributes {
  public static allStandardAttributes(additionalCustomAttributes: string[] = []): IUserPoolClientAttributes {
    return {
      email: true,
      phone: true,
      // ... all other standard attributes
      custom:attribute1: true,
      custom:attribute2: true,
      // any other custom attribute specified in `additionalCustomAttributes`
    }
  }
}
```

The usage will be:
```ts
const readAttributes = UserPoolClientAttributes.allStandard(['attr1', 'attr2']);
const writeAttributes = UserPoolClientAttributes.allStandard();
const pool = new UserPool(this, 'Pool');
pool.addClient('webapp-client', {
  readAttributes,
  writeAttributes,
});
```

One concern about extending `RequiredAttributes` is that its docstrings are not appropriate for the use in app client. (e.g. "Whether the user's postal address is a required attribute.") 
https://github.com/aws/aws-cdk/blob/40fa93a22ffbdf18b0563d1cef63bbf5814dcc3f/packages/%40aws-cdk/aws-cognito/lib/user-pool-attr.ts#L1-L23

Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:206]: > One concern about extending RequiredAttributes is that its docstrings are not appropriate for the use in app client.

We could modify the doc strings so that they are a bit more generic such as 'the user's postal address'. 

For the rest of your code, it's quite difficult to discuss this and comment on your code. Do you mind making these changes and then I can comment on whether they fit?
The important thing to consider is the user experience when a user uses this library. It needs to be nice and slick.

**Comment by tom139** [comment]
Sorry for making you wait so long. 🙏🏽

I started from scratch with the approach suggested by @nija-at inspired by `AccessLogFormat`. The main change is the addition of the `AttributeSet` class that represent a set of attributes (both standard and custom).

Naming of the methods may be misleading, but I couldn't think anything better and am open to suggestion.

Usage is:
```ts
const pool = new cognito.UserPool(this, 'Pool');
pool.addClient('app-client', {
  // ...
  readAttributes: AttributeSet.allStandard(['custom:my_attribute_1', 'custom:my_attribute_2']),
  writeAttributes: AttributeSet.from({name: true, locale: true}, ['custom:my_attribute_1']),
});
```

**Comment by nija-at** [comment]
Hey @tom139 -

I'm going to be on vacation for the next couple of weeks for the holidays. I'll take a look at this once I'm back.
Happy holidays!

**Comment by tom139** [comment]
Have a nice time and stay safe! :)

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:280]:
```suggestion
   * @default - all standard and custom attributes
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:287]:
```suggestion
   * @default - all standard and custom attributes
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:280]: Or should it be just the standard attributes?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:348]: Curious: why are preferred_email and preferred_number_number special? 
They don't feature as standard attributes in the cognito docs. why are they considered as standard?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:477]: Perhaps this would be a better API -


```ts
AttributeSet.withStandardAttributes(...).withCustomAttributes(...)

AttributeSet.withCustomAttributes(...).withStandardAttributes(...)

AttributeSet.empty().withCustomAttributes(...)

(new AttributeSet()).withStandardAttributes(...)
```

File: packages/@aws-cdk/aws-cognito/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:585]: Given that we know that it is custom, we could avoid the 'custom' prefix?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:350]: There are two sets of standard attributes now (this and the `StandardAttributes` interface).

Either we should consolidate them into one place, or at least, have some kind of test that ensures that they stay consistent.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:470]: How about naming this `ClientAttributes` or `UserPoolClientAttributes` ?
This makes it clear that these are attributes to be associated with a user pool client.

**Review by nija-at — DISMISSED** [review]
Thanks for updating this PR. Some questions, comments and suggestions below.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:470]: Good point, will rename it to `ClientAttributes`.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:477]: Totally agree with you. I replaced the old API with a new one that looks like:
```ts
// create write attributes from an empty set
const writeAttributes = ClientAttributes.empty()
  .withStandardAttributes({name: true, email: true})
  .withCustomAttributes(['favouritePizza']);
// read attributes are usually more than the writable ones, so we can start from those
const clientReadAttributes = clientWriteAttributes
  .withStandardAttributes({emailVerified: true})
  .withCustomAttributes(['pointsEarned']);
```

File: packages/@aws-cdk/aws-cognito/README.md
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/README.md:585]: Sure… much cleaner that way.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:348]: My bad. It is not `preferred_email` and `preferred_phone_number` to be special, it is `email_verified` and `phone_number_verified`.

These are _special_ because are not directly mentioned in the list at [the linked page](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html) but are present in the cited [OpenID Connect Specification](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) and are in fact present among the other standard attributes.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (tom139) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:475]: I think there could be place for 2 more static methods, for making a `ClientAttributes` instance:
 - with all standard attributes
 - with all standard attributes except for `emailVerified` and `phoneNumberVerified`.

This could be something like:
```ts
let readAttributes = ClientAttributes.allStandardAttributes();
let writeAttributes = ClientAttributes.allUserStandardAttributes(); // I don't like this method's name
```

You think it is worth it or the user should go fully verbose:
```ts
let readAttributes = ClientAttributes.empty().withStandardAttributes({
      address: true,
      birthdate: true,
      emailVerified: true,
      familyName: true,
      fullname: true,
      gender: true,
      givenName: true,
      lastUpdateTime: true,
      locale: true,
      middleName: true,
      nickname: true,
      phoneNumberVerified: true,
      preferredUsername: true,
      profilePage: true,
      profilePicture: true,
      timezone: true,
      website: true,
    });
let writeAttributes = ClientAttributes.empty().withStandardAttributes({
      address: true,
      birthdate: true,
      email: true,
      familyName: true,
      fullname: true,
      gender: true,
      givenName: true,
      lastUpdateTime: true,
      locale: true,
      middleName: true,
      nickname: true,
      phoneNumber: true,
      preferredUsername: true,
      profilePage: true,
      profilePicture: true,
      timezone: true,
      website: true,
    });
```

**Review by tom139 — COMMENTED** [review]
I think I addressed everything except setting up a test for detecting drifts between `StandardAttributes` and `StandardAttributesMask`.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts]: > are in fact present among the other standard attributes

What do you mean by this? Were you able to confirm that they are actually supported by Cognito, but just missing in the documentation?

If that's the case, we should just add these to `StandardAttributes`.

Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts]: I have conducted some tests and I confirm [Configuring User Pool](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html) is missing two attributes: `email_verified` and `phone_number_verified`.

Even though it states:

> You get a set of default attributes, called "standard attributes," with all user pools. You can also add custom attributes to your user pool definition in the AWS Management Console. This topic describes those attributes in detail and gives you tips on how to set up your user pool.

and does not list them under standard attributes, the following call:

```bash
❯ aws cognito-idp create-user-pool --cli-input-json file://userpool.json
❯ cat userpool.json 
{
    "PoolName": "test",
    "Policies": {
        "PasswordPolicy": {
            "MinimumLength": 10,
            "RequireUppercase": true,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": true,
            "TemporaryPasswordValidityDays": 10
        }
    },
    "AutoVerifiedAttributes": [
        "email"
    ],
    "UsernameAttributes": [
        "email"
    ],
    "Schema": [
        {
            "Name": "email",
            "AttributeDataType": "String",
            "Mutable": true,
            "Required": true,
            "StringAttributeConstraints": {
                "MinLength": "5",
                "MaxLength": "2048"
            }
        },
        {
            "Name": "email_verified",
            "AttributeDataType": "Boolean",
            "Mutable": true,
            "Required": true
        }
    ],
    "UsernameConfiguration": {
        "CaseSensitive": true
    }
}
```
works as expected, setting both `email` and `email_verified` as _required_ and _mutable_ attributes.

The UI Console has the same problem as it does not show `(phone_number|email)_verified` in the attribute list:
image

_I filed feedbacks for the documentation and the web ui pages._

That said, I will add those 2 attributes to the `StandardAttributes` interface, so `StandardAttributes` and `StandardAttributesMask` will have the same properties.

File: packages/@aws-cdk/aws-cognito/test/user-pool-attr.test.ts
Comment (tom139) [inline packages/@aws-cdk/aws-cognito/test/user-pool-attr.test.ts:289]: This method is meant to check whether all attributes in `StandardAttributeNames` are supported in `ClientAttributed` (i.e. are present in `StandardAttributesMask`).

@nija-at were you thinking to something like this in https://github.com/aws/aws-cdk/pull/7607/files?file-filters%5B%5D=.json&file-filters%5B%5D=.ts#r554988883 ?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:363]: I've added a test case checking whether all attributes in `StandardAttributeNames` are supported in `ClientAttributed` (i.e. are present in `StandardAttributesMask`).
see https://github.com/tom139/aws-cdk/blob/6b6630caa6dfa88a47960e8bcd0d211f46b6d521/packages/%40aws-cdk/aws-cognito/test/user-pool-attr.test.ts#L270-L289

Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:363]: Nice!

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:488]: Given that `false` is the default, let's start with fully verbose. 
We can add more static methods in the future as use cases arise.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts:488]: Should we have a `ClientAttributes.empty()` method or just use the constructor `new ClientAttributes()` like any other class?

File: packages/@aws-cdk/aws-cognito/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:596]: Slight tweaks to the documentation.

````suggestion
The default behaviour is to allow read and write permissions on all attributes. The following code shows how this can be configured for a client.

```ts
const pool = new cognito.UserPool(this, 'Pool');

const clientWriteAttributes = ClientAttributes.empty()
  .withStandardAttributes({name: true, email: true})
  .withCustomAttributes(['favouritePizza']);

const clientReadAttributes = clientWriteAttributes
  .withStandardAttributes({emailVerified: true})
  .withCustomAttributes(['pointsEarned']);
pool.addClient('app-client', {
  // ...
  readAttributes: clientReadAttributes,
  writeAttributes: clientWriteAttributes,
});
```
````

File: packages/@aws-cdk/aws-cognito/test/user-pool-attr.test.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/test/user-pool-attr.test.ts:272]: Nice! Can you update this test to also do the reverse? i.e., changes in StandardAttributesMask are reflected in StandardAttributeNames

**Review by nija-at — CHANGES_REQUESTED** [review]
Looks closer. A few more comments. Hopefully the last round of review before merging 😊

File: packages/@aws-cdk/aws-cognito/test/user-pool-attr.test.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/test/user-pool-attr.test.ts:272]: The problem with this is that `StandardAttributesMask` is an interface: this means the interface is erased at runtime and I have no idea how to iterate on its properties and detect changes. The only way I see is parsing directly the `lib/user-pool-attr.ts` or the `lib/user-pool-attr.d.ts`. Do you have any hint on how to proceed?

On the other end, the problem arise only if there is a new property added to the `StandardAttributesMask` but no test is setup to check if it is added to the `(new ClientAttributes).attributes()` array.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts
Comment (tom139) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-attr.ts]: Much nicer 👍🏽 I did the refactor.

**Review by tom139 — COMMENTED** [review]
Thank you for the time spent reviewing this!

**Comment by nija-at** [comment]
> am I supposed to hit "Resolve Conversation" when I think I addressed the issue

Clicking 'resolve conversation' is fine once you've addressed it 😊

**Review by nija-at — APPROVED** [review]

**Comment by russelldavis** [comment]
@tom139 @nija-at this is great, thank you! One question: why does `withStandardAttributes` take in an object like `{name: true, email: true}`, instead of a simple list like `["name", "email"]`? It's also inconsistent with `withCustomAttributes`, which takes in a list of parameters.

**Comment by russelldavis** [comment]
Also, regarding https://github.com/aws/aws-cdk/pull/7607#discussion_r556176241, I'd like to express interest in the additional static method you discussed, which would prepopulate all of the standard attributes.

A common use case (I think) is to allow write access to everything except for a small number of custom attributes. To do this now, I'll have to manually list all of the standard attributes, and keep my manual list up to date if those ever change.

Thanks again.

**Comment by russelldavis** [comment]
To follow up on that last comment, here's the code I ended up writing:

```js
const standardWriteAttributes = new ClientAttributes().withStandardAttributes({
  address: true,
  birthdate: true,
  email: true,
  familyName: true,
  gender: true,
  givenName: true,
  locale: true,
  middleName: true,
  fullname: true,
  nickname: true,
  phoneNumber: true,
  profilePicture: true,
  preferredUsername: true,
  profilePage: true,
  timezone: true,
  lastUpdateTime: true,
  website: true,
});

const standardReadAttributes = standardWriteAttributes.withStandardAttributes({
  emailVerified: true,
  phoneNumberVerified: true,
});

new UserPoolClient(this, "UserPoolClient", {
  // Allow everything except for writing custom attributes.
  readAttributes: standardReadAttributes.withCustomAttributes(...Object.keys(customAttributes)),
  writeAttributes: standardWriteAttributes,
});

```

Would be great if those lists were exposed, perhaps via `withAllStandardReadAttributes()` and `withAllStandardWriteAttributes()` functions.

Note that they need to be separate lists, because `emailVerified` and `phoneNumberVerified` cannot be set as writable. I learned this the hard way when `cdk deploy` gave me an error of "Invalid write attributes specified" without telling me which ones were invalid (another reason having these predefined would be helpful.)

**Comment by tom139** [comment]
Hi, thank you for your feedback! I totally agree with you. What do you think should be the best way to implement this?

#### Having two _all attributes_ methods
as you suggested: one for Writing and one for Reading attributes.
`.withAllStandardReadAttributes()` for all attributes,
`.withAllStandardWriteAttributes()` for all attributes except `*Verified`.

Usage:
```ts
const readAttr = (new ClientAttributes()).withAllStandardReadAttributes();
const writeAttr = (new ClientAttributes()).withAllStandardWriteAttributes();
```

#### Having only one method
`withAllStandardAttributes()` that adds all standard attributes and adding a check for removing `*Verified` attribute at _UserPoolClient_ creation time

Usage:
```ts
const readAttr = (new ClientAttributes()).withAllStandardAttributes();
const writeAttr = (new ClientAttributes()).withAllStandardAttributes();
```

---

I think the latter would be is easier to use, what do you think?

**Comment by tom139** [comment]
Sorry, I forgot to write about different design choices for _withCustomAttributes()_ vs _withStandardAttributes()`.

Having an interface for the standard attributes allow the user to easily find the available values through code completion; and in my opinion it is more valuable than consistency among `with*Attributes()` methods.

**Comment by russelldavis** [comment]
> What do you think should be the best way to implement this?

I like the idea of having as few methods as possible, but in this case I think keeping them separate is best. Automatically removing the `*Verified` attributes could cause some confusion — a developer would reasonably assume those attributes are writable if they put them in `writeAttributes` and everything succeeds, but they'd be wrong. "Explicit is better than implicit."

**Comment by russelldavis** [comment]
I'm with you about the value of code completion. Fortunately there's a way to have that without sacrificing consistency. For example:

```typescript
type StandardAttribute = "name" | "email";
function withStandardAttributes(...attributes: StandardAttribute[]);
```

When you type `withStandardAttributes("` in VSCode, it will give you code completion, and show an error if you type a value not in the list.

**Comment by tom139** [comment]
Yeah, it is very cool! Unfortunately we must ensure that JSII can synthesise this information in other languages too… I don't think this will work outside of Typescript. 😔

@nija-at could you help us on that?

**Pull request merged**
